Doc/library/sqlite3.rst

```diff
@@ -47,84 +47,148 @@ This document includes four main sections:
+   - if needed, use double quotes for table and column names
+
 .. _sqlite3-tutorial:
 
 Tutorial
 --------
 
-To use the module, start by creating a :class:`Connection` object that
-represents the database.  Here the data will be stored in the
-:file:`example.db` file::
+In this tutorial, you will learn the basics of the :mod:`!sqlite3` API
+by creating an on-disk movie database :file:`tutorial.db`
+an filling it with meta-data about Monty Python movies.
+A fundamental understanding of database concepts,
+like `transactions`_ and `cursors`_, is assumed.
+
+To start working with an SQLite database,
+you must first open a connection to it.
+Call the :func:`sqlite3.connect` function
+to create a connection to the database :file:`tutorial.db`::
 
    import sqlite3
-   con = sqlite3.connect('example.db')
+   con = sqlite3.connect("tutorial.db")
 
-The special path name ``:memory:`` can be provided to create a temporary
-database in RAM.
+SQLite will implicitly create an empty database file :file:`tutorial.db`
+in the current working directory if it does not exist.
+The returned :class:`Connection` object ``con``
+represents the connection to the on-disk database.
 
-Once a :class:`Connection` has been established, create a :class:`Cursor` object
-and call its :meth:`~Cursor.execute` method to perform SQL commands::
+In order to execute SQL statements and fetch results from SQL queries,
+use a database cursor.
+Call :meth:`~Connection.cursor` on ``con`` to create
+:class:`Cursor` ``cur``::
 
    cur = con.cursor()
 
-   # Create table
-   cur.execute('''CREATE TABLE stocks
-                  (date text, trans text, symbol text, qty real, price real)''')
+Now, create a database table ``movie`` with columns for title,
+year released, and `IMDB`_ review score.
+The `flexible typing`_ feature of SQLite makes typenames optional;
+for simplicity, just use column names in the table declaration.
+Execute the SQL statement
+by calling :meth:`~Cursor.execute` on ``cur``::
 
-   # Insert a row of data
-   cur.execute("INSERT INTO stocks VALUES ('2006-01-05','BUY','RHAT',100,35.14)")
+   cur.execute("CREATE TABLE Movie(Title, Year, Score)")
 
-   # Save (commit) the changes
-   con.commit()
+Verify that the table has been created by running a query against it.
+Execute the query by calling :meth:`~Cursor.execute` on ``cur``,
+store the result in a variable ``res``,
+and call :meth:`~Cursor.fetchone` on ``res`` to fetch the first
+(and only) row from the query::
 
-   # We can also close the connection if we are done with it.
-   # Just be sure any changes have been committed or they will be lost.
-   con.close()
+   >>> res = cur.execute("SELECT count(rowid) FROM Movie")
+   >>> res.fetchone()
+   (0,)
 
-The saved data is persistent: it can be reloaded in a subsequent session even
-after restarting the Python interpreter::
+The result is a one-item :class:`tuple`:
+one row, with one column.
+It correctly shows a row count of zero.
```

Comment:
This part is a bit confusing.  IIUC it tries to verify the existence of the table indirectly, by trying to read the content of the table and checking that no error is raised (?).  However it's not obvious that a row count of `0` (which seems to indicate lack of result/emptiness) means that the table exist.

Isn't there a query that can be used to list the tables (so that we can check if our `Movie` table is in there), or alternatively a query that returns true if the table exists?

Reply:
Hm, yeah. These things are obvious to me, but I understand that they might not be to everyone else. There are several solutions:

Once could prove that no table existed by first running a query and verifying we get an exception. However, I don't think that makes for a good tutorial.

We could query the SQLite master table, but that would be delving into SQLite internals, which would probably require expanding the tutorial with some sentences about the [sqlite_schema table](https://www.sqlite.org/schematab.html). I'm not sure if that makes for a better tutorial either.

Maybe the best thing to do, is just to mention that we expect a row count of zero, since there is now data in the table, and that we'd get an exception if the table was not created.

Reply:
I tried adding a variant of querying the schema; I'm not sure it worked out, though.

Reply:
Seems to make intuitive sense to me, for what its worth

Reply:
FWIW I think it's important to mention `sqlite_schema`.  Whenever I came across a DB the first thing that I want to do is printing a list of tables to get an idea of what's inside the DB, and every DB does that in a different way and I always have to google how to do it.

It might even be a good idea adding an "how to see the list of tables/columns".

Reply:
> (unresolving so the comment doesn't get lost)

Thanks!

> FWIW I think it's important to mention sqlite_schema. Whenever I came across a DB the first thing that I want to do is printing a list of tables to get an idea of what's inside the DB, and every DB does that in a different way and I always have to google how to do it.

That is a good point. There are some basic SQLite tricks that is worth knowing about. Perhaps we could put those into a howto? One of the most useful is the `pragma compile_options` query, from 3.12, even more available:

```
$ cd cpython.git
# configure && make
$ ./python -m sqlite3 :memory: "pragma compile_options" 
```

Reply:
Actually, mentioning the new CLI could be worth it for the >= 3.12 tutorial.